samza-core/src/main/java/org/apache/samza/system/inmemory/InMemoryManager.java

```diff
@@ -75,6 +75,19 @@ void put(SystemStreamPartition ssp, Object key, Object message) {
         .add(messageEnvelope);
   }
 
+  /**
+   * Handles produce request from {@link InMemorySystemProducer} for case where a job has a custom IME and
+   * populates the underlying message queue with the IME.
+   *
+   * @param ssp
+   * @param envelope
+   */
+  void put(SystemStreamPartition ssp, IncomingMessageEnvelope envelope) {
+    List<IncomingMessageEnvelope> messages = bufferedMessages.get(ssp);
+    bufferedMessages.get(ssp).add(envelope);
```

Comment:
Need to ensure the offset inserted into the queue is increasing and equals to last message `offset + 1` otherwise, poll logic will be impacted